File: src/main/java/com/clearcont/clearcontapp/service/ControleService.java
package com.clearcont.clearcontapp.service;

import com.clearcont.clearcontapp.model.Balancete;
import com.clearcont.clearcontapp.model.ComposicaoLancamentosContabeis;
import com.clearcont.clearcontapp.model.Controle;
import jakarta.transaction.Transactional;
import org.jetbrains.annotations.NotNull;
import org.springframework.stereotype.Service;

import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;

@Service
@Transactional
public class ControleService {
    private final BalanceteService balanceteService;

    public ControleService(BalanceteService balanceteService) {
        this.balanceteService = balanceteService;
    }

    public @NotNull List<Controle> getAllByMonthAndCompanyID(Long id, String month, Integer year) {
        List<Balancete> byCompanyAndPeriod = balanceteService.getByCompanyAndPeriod(id, month, year);
        List<Controle> controles = new ArrayList<>();
        byCompanyAndPeriod.forEach(
                balancete -> {
                    if (!balancete.getComposicaoLancamentosContabeisList().isEmpty())
                        controles.add(
                                new Controle(
                                        null,
                                        null,
                                        balancete.getClassificacao(),
                                        balancete.getNomeConta(),
                                        balancete.getDoubleTotalBalancete(),
                                        balancete.getComposicaoLancamentosContabeisList().stream().mapToDouble(
                                                ComposicaoLancamentosContabeis::getDoubleSaldoContabil).sum(),
                                        balancete.getComposicaoLancamentosContabeisList()
                                                .stream().mapToDouble(ComposicaoLancamentosContabeis::getDoubleSaldoContabil)
                                                .sum() - balancete.getDoubleTotalBalancete(),
                                        "ABERTO",
                                        "",
                                        false,
                                        false,
                                        LocalDate.now(),
                                        balancete.getComposicaoLancamentosContabeisList().getFirst().getResponsavel().getNome(),
                                        balancete.getEmpresa()
                                ));
                }
        );
        return controles;
    }
}

File: src/main/kotlin/br/com/clearcont/clearcontwebapp/service/ControleService.kt
package br.com.clearcont.clearcontwebapp.service

import br.com.clearcont.clearcontwebapp.models.Balancete
import br.com.clearcont.clearcontwebapp.models.ComposicaoLancamentosContabeis
import br.com.clearcont.clearcontwebapp.models.Controle
import jakarta.transaction.Transactional
import org.springframework.stereotype.Service
import java.time.LocalDate
import java.util.function.Consumer

@Service
@Transactional
class ControleService(private val balanceteService: BalanceteService) {
    fun getAllByMonthAndCompanyID(id: Long, month: String, year: Int): List<Controle> {
        val byCompanyAndPeriod = balanceteService.getByCompanyAndPeriod(id, month, year)
        val controles: MutableList<Controle> = ArrayList()
        byCompanyAndPeriod.forEach(
            Consumer { balancete: Balancete ->
                if (balancete.lancamentosContabeisList.isNotEmpty()) controles.add(
                    Controle(
                        null,
                        null,
                        balancete.classificacao,
                        balancete.nomeConta,
                        balancete.getTotalBalanceteDouble(),
                        balancete.lancamentosContabeisList.stream().mapToDouble(
                            ComposicaoLancamentosContabeis::saldoContabil).sum(),
                        balancete.lancamentosContabeisList.stream()
                            .mapToDouble(ComposicaoLancamentosContabeis::saldoContabil)
                            .sum() - balancete.getTotalBalanceteDouble(),
                        "ABERTO",
                        "",
                        composicaoPreenchida = false,
                        agingListadaPendencia = false,
                        dataCompetencia = LocalDate.now(),
                        nomeResponsavel = balancete.lancamentosContabeisList.first().responsavel.nome,
                        empresa = balancete.empresa
                    )
                )
            }
        )
        return controles
    }
}
